src/storage/src/hummock/shared_buffer/shared_buffer_manager.rs

```diff
@@ -107,88 +124,106 @@ impl SharedBufferManager {
-
-        // Atomically allocates space,
-        // since there could be concurrent Actors write to the shared buffer.
-        'retry_allocate: loop {
-            // flush shared buffer if there is no enough space
-            while threshold < current_size + batch_size {
-                log::debug!(
-                    "trigge flush: threshold {}, new_size {}",
-                    threshold,
-                    current_size + batch_size
-                );
-                self.flush().await?;
-                current_size = self.stats.shared_buffer_cur_size.load(Ordering::SeqCst);
-            }
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
 
-            let res = self.stats.shared_buffer_cur_size.compare_exchange(
-                current_size,
-                current_size + batch_size,
-                Ordering::SeqCst,
-                Ordering::Acquire,
-            );
-            match res {
-                Ok(_) => {
-                    assert!(current_size + batch_size <= threshold);
-                    break; // success
-                }
-                Err(old_val) => {
-                    current_size = old_val;
-                    continue 'retry_allocate;
-                }
-            }
-        }
-        Ok(())
+    pub fn scheduled_size(&self) -> usize {
+        self.scheduled_size.load(Ordering::Acquire)
     }
 
     /// Puts a write batch into shared buffer. The batch will be synced to S3 asynchronously.
     pub async fn write_batch(
         &self,
         batch: Vec<SharedBufferItem>,
         epoch: HummockEpoch,
-    ) -> HummockResult<u64> {
+    ) -> HummockResult<usize> {
         let batch = SharedBufferBatch::new(batch, epoch);
-        let size = batch.size;
+        let batch_size = batch.len();
 
-        self.allocate_space(size).await?;
+        // Stall writes if there is no quota.
+        loop {
+            let buffer_size = self.len();
+            if buffer_size == 0 || buffer_size + batch_size < self.capacity {
+                break;
+            }
+            self.uploader_notifier.notify_one();
+            self.release_notifier.notified().await;
+        }
 
-        self.shared_buffer
-            .write()
+        let mut inner = self.inner.write();
+        inner
+            .shared_buffer
             .entry(epoch)
             .or_insert(BTreeMap::new())
             .insert(batch.end_user_key().to_vec(), batch.clone());
+        inner.buffered_size += batch_size;
+        let size = inner.buffered_size - self.scheduled_size.load(Ordering::Acquire);
+        drop(inner);
+
+        // Queue batch in uploader.
         self.uploader_tx
             .send(SharedBufferUploaderItem::Batch(batch))
             .map_err(HummockError::shared_buffer_error)?;
-        Ok(size)
+
+        // Notify uploader to flush if not-uploading buffer size exceeds threshold.
+        if size > self.threshold {
+            self.flush();
```

Comment:
why not decide whether to flush in flush thread?

Reply:
Actually, I want to do a refactor on shared buffer uploading to reduce the code complexity later(e.g. maintain 2 btree map in both shared buffer and uploader). This PR only focus on concurrent issue in #1671 .